types/imperium/index.d.ts

```diff
@@ -0,0 +1,81 @@
+// Type definitions for imperium 2.0
+// Project: https://www.npmjs.org/package/imperium
+// Definitions by: Gaetan SENN <https://github.com/gaetansenn>
+// Definitions: https://github.com/psnider/DefinitelyTyped/imperium
+// TypeScript Version: 2.7
+
+/// <reference types="node" />
+
+import express = require('express');
+
+type GetAcl = (req: express.Request) => Promise<boolean> | Promise<object>;
+type Actions = string[] | string;
+type Context = Array<'params' | 'query' | 'headers' | 'body'>;
+// Can contain when key that is evaluated during route action
+interface RoleParams {
+    [key: string]: string;
+}
+
+export class Imperium {
+    constructor()
+
+    context: string[];
+    roles: Roles;
+
+    // Add new role with specific ImperiumGetAcl
+    role(roleName: string, getAcl?: GetAcl): Role;
+
+    // Check if user has role(s) act like as an OR
+    is(roleNames: string | string[]): Promise<express.RequestHandler>;
+
+    // Check if current user can do action(s)
+    can(actionS: string | string[] | Action | Action[]): Promise<express.RequestHandler>;
+
+    private addRole(roleName: string, getAcl: GetAcl): void;
+
+    evaluateRouteActions(req: express.Request, action: Action[], context: Context): Actions;
+
+    evaluateRouteAction(req: express.Request, expr: string, key: string, context: Context): string;
+
+    evaluateUserActions(req: express.Request, roles: Role[]): Promise<Action[]>;
+
+    evaluateUserAction(action: RoleParams, context: { [key: string]: string[] }): { [key: string]: string[] };
+}
+
+interface Roles {
+    [key: string]: RoleActions;
+}
+
+interface RoleActions {
+    actions: Action[];
+    getAcl?: GetAcl;
+}
+
+interface Action {
+    action: string;
+    [key: string]: string;
+}
+
+export class Role {
+    constructor(imperium: Imperium, roleName: string)
+
+    // Imperium instance to retreive child role
+    imperium: Imperium;
+    // Role name
+    roleName: string;
+    // Contain all the actions for this specific role
+    role: RoleActions;
+
+    /* Add action with specific params */
+    can(action: string, params: RoleParams): Role;
+
+    /* Get actions of childRoleName and replace params */
+    is(childRoleName: string, params: RoleParams): Role;
+}
+
+export class UnauthorizedError extends Error {
+    constructor(message: string, status: number, context: any)
+}
+
+declare const init: Imperium;
+export default init;
```

Comment:
Given that the underlying library doesn't have an `export default`, your declaration file should not either.

Unfortunately, given the shape of the library, the best way to handle this is to copy the methods on the class `Imperium` to the top level and export them as well:

```
export const context: string[];
export const roles: Roles;

// Add new role with specific ImperiumGetAcl
export function role(roleName: string, getAcl?: GetAcl): Role;

// Check if user has role(s) act like as an OR
export function is(roleNames: string | string[]): Promise<express.RequestHandler>;

// Check if current user can do action(s)
export function can(actionS: string | string[] | Action | Action[]): Promise<express.RequestHandler>;

export function evaluateRouteActions(req: express.Request, action: Action[], context: Context): Actions;

export function evaluateRouteAction(req: express.Request, expr: string, key: string, context: Context): string;

export function evaluateUserActions(req: express.Request, roles: Role[]): Promise<Action[]>;

export function evaluateUserAction(action: RoleParams, context: { [key: string]: string[] }): { [key: string]: string[] };

export class Imperium {
    constructor()

    context: string[];
    roles: Roles;

    // Add new role with specific ImperiumGetAcl
    role(roleName: string, getAcl?: GetAcl): Role;
```
and so on.

Given this, you can then remove lines 80 and 81.

Reply:
Hi @uniqueiniquity Actualy the library is currently exporting an Imperium instance by default.

From our library at the line 186 we are doing an export default
https://github.com/mono-js/imperium/blob/master/lib/index.js

```
module.exports = new Imperium()
```

Maybe I dont' have understand you comment but from the imperium-test.ts, we are using the Imperium instance from : 
```
import imperium from 'imperium';
```

Regards,

Reply:
Hi there @gaetansenn, that's not an `export default` which is a relatively new syntax introduced in ECMAScript 2015. That's instead an assignment to `module.exports` which only works on certain types of module formats that predate ECMAScript 2015 (e.g. CommonJS, AMD, etc.).

The fact that you're able to import it as a default import (i.e. `import imperium from 'imperium'`) is thanks to the interop strategy that you're getting potentially from Webpack, Parcel, or SystemJS, **or**, you're using TypeScript's `esModuleInterop`.

So our suggestion is

1. Be faithful to the original module format: because you've used `module.exports = ...`, you should use the analogous TypeScript construct: `export = ...`
2. If this library's userbase commonly uses default imports, you can use that in your tests using the `esModuleInterop` flag.

Reply:
We are using CommonJS as this library is used for nodeJS applications. I updated the tsconfig.js as you advise from your second suggestion.

Thanks a lot for you feedback.

Regards,